### PR TITLE
Access to CollectionType options in AdminType

### PR DESCRIPTION
**Pull request by VincentLanglet**
This is related to https://github.com/sonata-project/SonataAdminBundle/issues/6426#issuecomment-702330648

I need to know the CollectionType value of `by_reference` in the AdminType.

The options are passed to the AdminType here:
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L172-L180

So I need to change the signature of `getDefaultOptions` to allow passing the `formOptions`.

But then I have two solutions: 
- Option 1: Adding a new `symfony_form_options` option to the AdminType, then in the FormContractor I will add 
```
$options['symfony_form_options'] => $formOptions
```
And then I access to `by_reference` with `$options['symfony_form_options']['by_reference'] ?? true`.

- Option 2: Using the `by_reference` option of the AdminType, then in the FormContractor I will add 
```
if (isset($formOption['by_reference'])) {
    $options['by_reference'] => $formOptions['by_reference'];
}
```
And then I access to `by_reference` in the AdminType with `$options['by_reference']`.

WDYT @sonata-project/contributors ?

In this PR I have implemented the code for the solution 1, but passing all the options seems overkill to me and I think the solution 2 more elegant.

### CODE REVIEW

**Comment by wbloszyk** [comment]
IMO When you exnteds `AbstractType` then you should have symfony options directly in $options. Also in case you can add own options to `$options` or `$options['sonata'] and never vice versa (never $options['symfony']). This will make usage more logic.

I vote for `option 2` :laughing:

File: src/Builder/FormContractorInterface.php
Comment (core23) [inline src/Builder/FormContractorInterface.php:45]: Can you add `@phpstan-param` information for this new parameter

Comment (VincentLanglet) [reply src/Builder/FormContractorInterface.php:45]: I tried, and phpstan is giving an error.

File: src/Form/Type/AdminType.php
Comment (VincentLanglet) [inline src/Form/Type/AdminType.php:115]: Not sure If it's better to do nothing if `by_reference` is not set ? Or adding a default value to true ?

**Comment by VincentLanglet** [comment]
Please review this bug fix @sonata-project/contributors

File: composer.json
Comment (jordisala1991) [inline composer.json:71]: do we need to add this conflict here?

File: src/Form/Type/AdminType.php
Comment (jordisala1991) [reply src/Form/Type/AdminType.php:115]: What could happen if we do nothing?

File: composer.json
Comment (VincentLanglet) [reply composer.json:71]: If you use SonataAdmin 3.78 with an old persistence bundle, the `by_reference` option won't be passed.
It will lead to a behavior different than the actual one in 3.77.
People will have this bug (https://github.com/sonata-project/SonataAdminBundle/issues/6166) again.

I can remove the conflict if you think that's not a good reason.

File: src/Form/Type/AdminType.php
Comment (VincentLanglet) [reply src/Form/Type/AdminType.php:115]: If we do nothing it will be the same than some older SonataAdmin versions, parentObject won't be set to the children one during the form display so people can't use it to display some conditional fields.

1) No set: Behavior in the beginning of 2020
2) AddInstance: Behavior until the summer of 2020 => Allow conditionnal checks, but create #6166
3) SetObject: Behavior until now => Fix 6166, but create #6426

If I do nothing when the option is not set, it could lead to a regression for people with an older persistence bundle (that's why I added conflict) or custom persistence bundle.
But I agree that doing nothing is the more natural. Maybe should I add a NEXT MAJOR comment ? @jordisala1991

File: composer.json
Comment (VincentLanglet) [reply composer.json:71]: https://github.com/sonata-project/SonataAdminBundle/pull/6438#discussion_r507509807

File: src/Form/Type/AdminType.php
Comment (jordisala1991) [reply src/Form/Type/AdminType.php:115]: What about:

1. we do not add conflict
2. we do here something to avoid changing default behavior for older persistence bundles (is there a difference here between older persisntence and new persistence releases)
3. maybe we add next major to clear this default thing? ( not sure about this one)

Comment (VincentLanglet) [reply src/Form/Type/AdminType.php:115]: I made more tests. The option is always set, so I removed the isset, but

Previously the persistence bundle didn't pass the option
Now: 
- If the option is set on the CollectionType, the persistence pass it
- If the option is not set, the persistence doesn't pass it BUT it used the default value (true) which is the same than the AdminType.

If someone use this code, it will lead to a behavior change if the option was false and not passed (because of an older persistence bundle version).
If I add something to say that the value will default to `false` if not passed, it won't fix the bug.

IMHO
- either we add conflict
- either we accept the minor change in the behavior (https://github.com/sonata-project/SonataAdminBundle/issues/6166 will appear again) and we recommend to update persistence bundle.

Comment (jordisala1991) [reply src/Form/Type/AdminType.php]: Then lets leave the conflict. I tried to avoid it because it is hard to know how and when we can remove this conflict, probably on 4.x. But it could be better IMO to add that restriction here: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/9e547e976f6d82bf2c40e33a1c637b9826cac60d/composer.json#L32

Comment (VincentLanglet) [reply src/Form/Type/AdminType.php]: I can't add the conflict on persistence bundle, since I want to block SontaAdmin 3.78+ with SonataDoctrineORM 3.23-

I leave the conflict. The PR is RTM.

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Tests are failing now

**Comment by VincentLanglet** [comment]
Fixed @jordisala1991

File: tests/Form/Type/AdminTypeTest.php
Comment (phansys) [inline tests/Form/Type/AdminTypeTest.php:214]: I think we can use `exactly()` instead of `atLeastOnce()` here, since this test is not using variable arguments.

Comment (VincentLanglet) [reply tests/Form/Type/AdminTypeTest.php]: Done

**Review by wbloszyk — APPROVED** [review]

File: src/Form/FormMapper.php
Comment (franmomu) [inline src/Form/FormMapper.php:134]: Probably I'm missing something, but why do we need to pass the form options to the builder (persistence bundles) if we already have them already here?

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: Because they are transforme.

From
```
$options = ['by_reference' => true];
```
To
``` 
$options = [`type_options` => ['by_reference' => true]]
```
(And multiple others modifications)

And the options passed are not always the same ; the depends on the `type`, etc

Comment (phansys) [reply src/Form/FormMapper.php:134]: I remember some unintended behavior regarding these options (#6078).

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: Not sure to understand. ^^'
Is it an answer to franmomu ? A request change to me ?

Comment (phansys) [reply src/Form/FormMapper.php:134]: Sorry. The comment is regarding the modifications you've mentioned about the options.

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: I still don't understand why it's an issue.

Previously it was
```
$options = array_replace_recursive(
                $this->builder->getDefaultOptions($type, $fieldDescription),
                $options
)
```

So $options was updated from
```
$options = ['foo' => 'bar' , 'by_reference' => false, ... ]
```
To
```
$options = ['foo' => 'bar' , 'by_reference' => false, ..., 'type' => AdminType::class, 'type_options' => [ ... ]. ]
```

Now, it's 
```
$options = array_replace_recursive(
                $this->builder->getDefaultOptions($type, $fieldDescription, $options),
                $options
)
```

So $options is updated from
```
$options = ['foo' => 'bar' , 'by_reference' => false, ... ]
```
To
```
$options = ['foo' => 'bar' , 'by_reference' => false, ..., 'type' => AdminType::class, 'type_options' => [ ..., 'by_reference' => false ]. ]
```

Franmomu was asking why the options is passed to DoctrineORM when it could be done in SonataAdmin.
But seems sometime the `getDefaultOptions` returns some `type_options`, sometimes there is not, I think it's better to do this in the persistence bundle (and it's already done).

Comment (phansys) [reply src/Form/FormMapper.php:134]: I'm not saying there is an issue with your changes. I'm pointing out that there is already some unexpected modifications in the options, related to the usage of `array_merge_recursive()` as mentioned in the linked PR https://github.com/sonata-project/SonataAdminBundle/pull/6078#issue-412703816.

Comment (franmomu) [reply src/Form/FormMapper.php:134]: What I meant is... is there a difference between this code and the modified one in persistence bundles and:

```php
$formOptions = $options;
$options = array_replace_recursive($this->builder->getDefaultOptions($type, $fieldDescription) ?? [], $formOptions);

if (isset($options['type_options') && isset($formOptions['by_reference'])) {
    $options['type_options']['by_reference'] = $formOptions['by_reference'];
}     
```

which means to leave persistence bundles as they were (less duplicated code and less to maintain) and not adding a conflict with them.

In fact looking at [`FormContractor:: getDefaultOptions`](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/711eba35e0bef87ab57fa5d4a8ee8b5f6457a62f/src/Builder/FormContractor.php#L99), looks that could be done in the admin bundle since the only persistence-related reference I see is to `ClassMetadata` and that could be abstracted in the `FieldDescriptionInterface`.

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: > What I meant is... is there a difference between this code and the modified one in persistence bundles and:
> 
> ```
> $formOptions = $options;
> $options = array_replace_recursive($this->builder->getDefaultOptions($type, $fieldDescription) ?? [], $formOptions);
> 
> if (isset($options['type_options') && isset($formOptions['by_reference'])) {
>     $options['type_options']['by_reference'] = $formOptions['by_reference'];
> }     
> ```

IMHO, since the persistence bundle are setting some options in `type_options`, everything should be done in it.
Maybe some will need to set the `by_reference` and some won't. We can't say ; only the persistence bundle will decide and we shouldn't guess in the SonataAdminBundle.

> In fact looking at FormContractor:: getDefaultOptions, looks that could be done in the admin bundle since the only persistence-related reference I see is to ClassMetadata and that could be abstracted in the FieldDescriptionInterface.

This PR is a patch for a bug existing for 1 month. Can we merge it ? 
A refacto to move all the logic from persistence bundle to admin bundle is certainly a good idea but it will take times.

**Comment by VincentLanglet** [comment]
@franmomu Can we merge this bugfix and decide after if we need and how to refactor getDefaultOptions ?

File: src/Form/FormMapper.php
Comment (phansys) [reply src/Form/FormMapper.php:134]: If the refactoring is the better choice we have, I think we should create an issue to track the required changes and update the `NEXT_MAJOR` comment in order to avoid pointing the goal to an approach that we consider slack.

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: > If the refactoring is the better choice we have, I think we should create an issue to track the required changes and update the `NEXT_MAJOR` comment in order to avoid pointing the goal to an approach that we consider slack.

If I understand correctly, @franmomu would like to deprecate `getDefaultOptions` since
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L99
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3.x/src/Builder/FormContractor.php#L97
are similar.

But if you look at other bundle, like 
https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/blob/2.x/src/Builder/FormContractor.php#L118
That's less similar. 

So I'm not sure the issue should be resolved.
We will have to live with similar code between ORM and MongoDB, but if we want to allow other persistence bundle, that's the right way to do it.

If @franmomu see things differently, I invite him to create the issue to explain the strategy.

But this debate should not block this PR.

**Comment by VincentLanglet** [comment]
Please review @sonata-project/contributors

File: src/Form/FormMapper.php
Comment (phansys) [reply src/Form/FormMapper.php:134]: The reasoning seems fine to me.
I just have a concern. Since we are adding a 3rd argument in order to fix this issue properly in the next major, shouldn't be better to provide the fix by using this new argument through `func_get_args()` and deprecate the case where it's considered required and not used?
I didn't see how this new argument will be used in the method, so in case my previous concern can not be done, I think at least we should leave the comment with the code fragment that must be uncommented in the next major.

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: I don't understand @phansys.

To me, the only NEXT_MAJOR comment needed is https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L101

The 3rd argument will always be used from now, so I don't understand why you want to deprecate it ?
This is a long term fix.

Comment (phansys) [reply src/Form/FormMapper.php:134]: >The 3rd argument will always be used from now, so I don't understand why you want to deprecate it ?

I'm not saying we should deprecate the argument, please re-read my comment:

>deprecate the case where it's considered required and not used

Comment (phansys) [reply src/Form/FormMapper.php:134]: >To me, the only NEXT_MAJOR comment needed is https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L101

Sorry, didn't see that the call to `func_get_args()` and the `NEXT_MAJOR` comment were already there before this PR.

I think then we should just trigger a deprecation if the `FormContractor::getDefaultOptions()` method is called without the 3rd argument. Do you think it could be done in this PR?

Comment (VincentLanglet) [reply src/Form/FormMapper.php:134]: The getDefaultOptions is implemented in persistence bundle, so I can't do it in this SonataAdmin PR.

But I can create PR on persistence bundle then.

**Review by phansys — APPROVED** [review]

**Pull request merged**
